### PR TITLE
Add rate limit avoidance to MeetupClient.

### PR DESCRIPTION
**Pull request by rbevers**
The existing client would hit rate limits if used for repeated Meetup calls in a short time frame. New logic added to MeetupClient will avoid it by thread sleeping until the rate limit time window is past.

### CODE REVIEW

File: meetup/api.py [hunk 1]
```diff
@@ -74,26 +81,84 @@ def invoke(self, meetup_method, params=None, method='GET'):
...
         try:
-            return json.loads(content)
+            self._capture_rate_limit(response)
+            return response.json()
         except:
             return None
 
     def _post(self, url, kwargs):
-        content = requests.post(url, data=kwargs).text
+        response = requests.post(url, data=kwargs).text
```
Comment (abendig) [inline meetup/api.py]: That leaves response as a string.

File: meetup/api.py [hunk 2]
```diff
@@ -74,26 +81,84 @@ def invoke(self, meetup_method, params=None, method='GET'):
...
         try:
-            return json.loads(content)
+            self._capture_rate_limit(response)
+            return response.json()
```
Comment (abendig) [inline meetup/api.py]: if response is a str it won't have a json() method

File: meetup/api.py [hunk 3]
```diff
@@ -74,26 +81,84 @@ def invoke(self, meetup_method, params=None, method='GET'):
...
         except:
             return None
```
Comment (abendig) [inline meetup/api.py:163]: ... and this will always be executed

**Review by abendig — CHANGES_REQUESTED** [review]
This breaks POST.

**Comment by rbevers** [comment]
Fix to the wrong response pattern. Ouch. :( Look better?

**Comment by abendig** [comment]
We need to add proper test coverage of the changes, before merging this out.

**Comment by rbevers** [comment]
Test coverage added in `test_api.py`. Runnable with:

    PYTHONPATH=$(pwd) python meetup/tests/test_api.py

**Comment by rbevers** [comment]
@abendig Is this sufficient test coverage?

File: meetup/api.py [hunk 4]
```diff
@@ -51,21 +57,22 @@ def invoke(self, meetup_method, params=None, method='GET'):
...
         response : dict
-
         """
         # TODO: rename invoke to http_response
 
         # get the parameters
         params = params.copy() if params is not None else {}
         params['key'] = self.api_key
-        params['page'] = 1000
+        if not params.has_key('page'):
```
Comment (abendig) [inline meetup/api.py]: _Slightly_ better: `if not "page" in params`

File: meetup/tests/test_api.py [hunk 5]
```diff
@@ -10,24 +10,149 @@
...
-MEETUP_KEY =  settings.MEETUP_KEY
+
+MEETUP_KEY = "abc123"
+
 
 # ########################################################################### #
 
 
-# TODO: implement tests
+class MeetupClientTests(unittest.TestCase):
```
Comment (abendig) [inline meetup/tests/test_api.py:30]: Thank you!

**Review by abendig — APPROVED** [review]
Let's ship!

**Comment by rbevers** [comment]
Ah, I forgot to merge. No merge button. Merging now.

**Comment by rbevers** [comment]
Oh, I am not a contributor on your repo. Can I merge the dependency to point to `rbevers/django-meetup` or do you want to merge my PR to your repo?

**Pull request merged**